util/py/scripts/mapfile_to_json.py

```diff
@@ -0,0 +1,339 @@
+                            func = m.group('function')
+                            if ('=' in func or 'ASSERT' in func or
+                                    'ALIGN' in func):
+                                continue
+                            if sections:
+                                sections[-1]['children'][-1]['symbols'].append(
+                                    {
+                                        'name': func,
+                                        'addr': addr,
+                                    })
+                        else:
+                            break
+                else:
+                    sections.append(obj)
+        self.sections = sections
+
+    def parse(self, filename):
+        """Parse the given mapfile."""
+        with open(filename) as f:
+            self.parse_memory_config(f)
+            self.parse_sections(f)
+
+    def by_memory_region(self):
+        """Return the mapfile arranged by memory region."""
+        region = {}
+        for sec in self.sections:
+            r = self.get_region(sec['addr'])
+            if r not in region:
+                region[r] = {
+                    'name': r,
+                    'addr': self.memory[r][0],
+                    'size': self.memory[r][1] - self.memory[r][0],
+                    'children': [],
+                    'symbols': [],
+                }
+            region[r]['children'].append(sec)
+        top = {
+            'name': 'all',
+            'children': list(region.values()),
+            'symbols': [],
+        }
+        return top
+
+    def sum_children(self, item):
+        """Sum up children and report discrepancies between sizes."""
+        name = item.get('name', '__unknown__')
+        size = item.get('size', 0)
+        children = item.get('children', [])
+        csz = 0
+        for c in children:
+            csz += self.sum_children(c)
+        if csz == 0:
+            csz = size
+        else:
+            item.pop('size', 0)
+            item['computed_size'] = csz
+        if size and csz != size:
+            logging.warning(f'warn: {name} size {size} != computed_size {csz}')
+        return csz
+
+
+def main(args):
+    m = Mapfile(regions=args.regions.split(','),
+                ignore_debug=args.ignore_debug)
+    m.parse(args.mapfile)
+    regions = m.by_memory_region()
+    m.sum_children(regions)
+    text = json.dumps(regions, indent=4)
+    if args.json:
+        with open(args.json, 'wt') as f:
+            f.write(text)
+    if args.html or args.view:
+        filename = args.html if args.html else os.path.basename(
+            args.mapfile).replace('.map', '') + '.html'
+        with open(filename, 'wt') as f:
+            f.write(D3_TEMPLATE.replace('@DATA@', text))
+        if args.view:
+            subprocess.call(['xdg-open', filename])
+    return 0
+
+
+if __name__ == '__main__':
+    args = parser.parse_args()
+    sys.exit(main(args))
+
+# vim: set ts=4 sts=4 sw=4 expandtab:
```

Comment:
?

Reply:
A vim modeline to override ${company}'s default python coding style.